### PR TITLE
[Inductor][CPP] Extract common functions to be reused in other CPP Template

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140786
* #140921
* #137975
* __->__ #141554

**Summary**
Extract common internal functions from GEMM Template into public function, so these functions can be reused by the  subsequent group GEMM template. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Review by jgong5 — COMMENTED** [review]
Could you please add function doc for the newly added functions.

**Comment by leslie-fang-intel** [comment]
> Could you please add function doc for the newly added functions.

Thanks for the comment. Adding the type hint and doc for these these new added functions.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 1]
```diff
@@ -430,7 +457,15 @@ def process_out_template_epilogues(
...
     reindexers: List[Optional[Callable[[List[Any]], List[Any]]]],
-):
+) -> tuple[
+    List[ir.IRNode],
+    ir.Buffer,
+    List[Optional[Callable[[List[Any]], List[Any]]]],
+    Union[ir.Buffer, ir.ReinterpretView],
+]:
+    """
+    Helper function to pre-process out template epilogues for code generation.
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_gemm_template.py:467]: Some comments:
1. It is more beneficial to explain why we need this function and give a high-level summary of what it does.
2. After clarify what it does, you will see it is cleaner to move the building of epilogues out of the function to perhaps a dedicated function.
3. Then, you can consider to give a better name to this function according to the function doc.
```suggestion
    The dimension and the indexing could be different between the GEMM output, i.e. `template_buffer`, which is
    2D with MxN) and the output from the template after epilogues, i.e. `Y`. In the GEMM template code,
    we are not aware of the dimension and the indexing of the epilogues and always work on 2D tiles according to
    the indexing of the GEMM output.
    In this function, we return a 2D buffer (`Y_2d`) according to GEMM output (reinterpreted from `Y` if needed) and
    build a reindexer that converts the indexing of `Y` into `Y_2d`.
```

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 2]
```diff
@@ -301,19 +307,36 @@ def transpose_w(W, trans_w):
...
             B = L.expand(B, (X.get_size()[0], B.get_size()[-1]))
         else:
             assert isinstance(B, torch.Tensor)
+            assert isinstance(X, torch.Tensor)
             B = B.expand(X.shape[0], B.shape[-1])
     return B
 
 
-def get_block_w(W, padded_n, block_n, k, n, micro_gemm):
+def get_block_w(
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_gemm_template.py:329]: Thanks for the comment and changed.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 3]
```diff
@@ -361,7 +384,11 @@ def get_block_w(W, padded_n, block_n, k, n, micro_gemm):
     return blocked_w
 
 
-def prune_tensors(input_nodes, new_input_nodes):
+def prune_tensors(input_nodes: List[ir.TensorBox], new_input_nodes: List[ir.TensorBox]):
+    """
+    Prune unused tensors since the GEMM Template use new packed weight.
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_gemm_template.py:389]: Thanks for the comment and changed.

File: torch/_inductor/codegen/cpp_gemm_template.py (hunk 1 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_gemm_template.py]: Thanks for the comment. Refactor the implementation by removing `epilogues` and rename this function to `gen_2d_view_of_epilogue_buf`.

**Review by jgong5 — APPROVED** [review]

**Pull request closed without merging**